### PR TITLE
[fix] nn.Embedding: allow changing the padding vector

### LINKED ISSUE #53368: Undocumented change of behavior for Embeddings in PyTorch 1.8
## 📚 Documentation

I'm filing this as a documentation problem but it could also be a bug if the documentation is what you want to be right.

Starting PyTorch 1.8, the `Embedding` layer will always output a vector of zeros for the `padding_idx` used to create it. Here is a short snippet to observe the behavior:

```python
import torch
import torch.nn as nn

embeds = nn.Embedding(5, 12, padding_idx=0) # Weights at 0 are initialized with zeros
embeds.weight.data[0] = torch.ones(12) # Put another value
embeds(torch.tensor([0]))
```

In Pytorch <= 1.7.1 this will return a tensor of ones, using the updated weights. Starting in PyTorch 1.8, this return a tensors of zeros.

The [documentation](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html#torch.nn.Embedding) still states that:

"""
With padding_idx set, the embedding vector at padding_idx is initialized to all zeros. However, note that this vector can be modified afterwards, e.g., using a customized initialization method, and thus changing the vector used to pad the output. The gradient for this vector from Embedding is always zero.
"""

I have no idea if the change was intended and the documentation was not updated accordingly, or if the change was not intended in which case there is a bug to fix. If you could tell me which one it is, it would be much appreciated!

In the first case I think this would warrant a clear warning in the release notes as it was not obvious to me from reading them and I had a lot of pretrained models not working the same way in 🤗 Transformers when upgrading to 1.8.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @brianjo @mruberry

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #53368

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -12366,6 +12366,14 @@ def test_embedding_padding_idx(self, device, dtype):
         self.assertEqual(output[0][2].sum(), 0)
         self.assertEqual(output[1][1].sum(), 0)
 
+        # change padding vector
+        padding_vector = torch.ones(20, dtype=dtype)
+        embedding = nn.Embedding(10, 20, padding_idx=2, sparse=True).to(device, dtype)
+        embedding.weight.data[2] = padding_vector
```
Comment (vadimkantorov) [inline test/test_nn.py:12372]: An idea: to add a `.padding_embedding` property on `nn.Embedding`.
It'd be more intuitive to do `embedding.padding_embedding.copy_(padding_vector)` than `embedding.weight.data[2] = padding_vector`? Hmm, although we couldn't make use of assignment operator in that case? ;/

Comment (albanD) [inline test/test_nn.py:12372]: nit: you should never use `.data`.
If you want to update the weights without the autograd tracking it, you should use `with torch.no_grad()`

File: test/test_nn.py [hunk 2]
```diff
@@ -12366,6 +12366,14 @@ def test_embedding_padding_idx(self, device, dtype):
         self.assertEqual(output[0][2].sum(), 0)
         self.assertEqual(output[1][1].sum(), 0)
 
+        # change padding vector
```
Comment (jbschlosser) [inline test/test_nn.py:12369]: In addition to this test, I think it'd be good to add to the `test_embedding_functional` test that compares functional and module forms.

In particular, the test should assert that, given the same `padding_idx` and `weight`, the module and functional outputs should *differ* when the embedding vector at `padding_idx` for the module is modified from its all-zero initialization.

The functional form should return zeros for `padding_idx` and the module form should return the modified embedding value.

File: torch/nn/functional.py [hunk 3]
```diff
@@ -1904,13 +1904,15 @@ def embedding(
     else:
         padding_idx = -1
     if max_norm is not None:
+        # Note [embedding_renorm contiguous]
```
Comment (jbschlosser) [inline torch/nn/functional.py:1907]: I can't comment on this line because it's not in the review, but for the `padding_idx` docs above (line 1852), I think the wording is strange.

"initialized to zeros" doesn't really make sense in the functional context where `weight` is passed in. Imo the wording should be changed to indicate that the output is padded with an all-zero vector wherever padding_idx is encountered.

**Review by jbschlosser — COMMENTED** [review]
Edit: After digging into this more (see my comment [here](https://github.com/pytorch/pytorch/issues/53368#issuecomment-793284794)), I think the behavior of both #46714 and #47184 should be reverted. Will review according to this.

**Review by jbschlosser — COMMENTED** [review]
Changes look correct to me! Verified that the behavior has been reverted to pre-1.8.0 semantics. Thanks for knocking this out :)

Suggested some doc changes, and it seems really important to get this right, given the amount of confusion the doc wording has caused in the past. @gchanan want to weigh in here on the wording?

File: torch/nn/functional.py [hunk 4]
```diff
@@ -1850,7 +1850,8 @@ def embedding(
         weight (Tensor): The embedding matrix with number of rows equal to the maximum possible index + 1,
             and number of columns equal to the embedding size
         padding_idx (int, optional): If given, pads the output with the embedding vector at :attr:`padding_idx`
-                                         (initialized to zeros) whenever it encounters the index.
+                                    whenever it encounters the index.
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/nn/functional.py:1853]: I'd prefer to update the docs seperately in a follow-up PR

**Review by jbschlosser — APPROVED** [review]
